opm/simulators/linalg/bda/openclSolverBackend.cpp

```diff
@@ -518,6 +506,49 @@ void openclSolverBackend<block_size>::initialize(int N_, int nnz_, int dim, doub
             d_toOrder = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * Nb);
         }
 
+        get_opencl_kernels();
+
+        prec->setKernels(ILU_apply1_k.get(), ILU_apply2_k.get(), ilu_decomp_k.get());
+
+    } catch (const cl::Error& error) {
+        std::ostringstream oss;
+        oss << "OpenCL Error: " << error.what() << "(" << error.err() << ")\n";
+        oss << getErrorString(error.err());
+        // rethrow exception
+        OPM_THROW(std::logic_error, oss.str());
+    } catch (const std::logic_error& error) {
+        // rethrow exception by OPM_THROW in the try{}, without this, a segfault occurs
+        throw error;
+    }
+
+    initialized = true;
+} // end initialize()
+
+template <unsigned int block_size>
+void openclSolverBackend<block_size>::get_opencl_kernels() {
+
+        cl::Program::Sources source(1, std::make_pair(axpy_s, strlen(axpy_s)));  // what does this '1' mean? cl::Program::Sources is of type 'std::vector<std::pair<const char*, long unsigned int> >'
```

Comment:
It means a vector with 1 (source) string and all strings are initialized with `axpy_s`. As `axpy_s` is a C string (or even a `std::string`, I don't see why the `make_pair` should be needed (actually I am not even sure how it works :flushed:).
Maybe kip the comment and use the following (untested as always, which might not be smart)?
```
cl::Program::Sources source;
source.emplace_back(axpy_s);
....
```

Reply:
`cl::Program::Sources` is defined as `typedef VECTOR_CLASS<std::pair<const char*, ::size_t> > Sources;`
The `1` is used in the opencl example in their header.

Reply:
That seems to depend on whichj cl*.hpp header we are using. The newer, improved, but unfortunately incompatible one seems to be cl2.hpp. That simplifies this construction with strings. Messy header business as you can only use one or the other. Let's stick with the old, but simplify the code.

The example they give is for old C++ version. With the newer ones you can simply do:
```
cl::Program::Sources source;
source.emplace_back(axpy_s, strlen(axpy_s));
...
```

I know the `1` is in the example, I just wanted to answer the comment. For multiple sources initializing the vector with size one seems strange.

Reply:
Concerning messy business:The two headers even define/use the same name for the include guard.

Reply:
I cleaned it up a bit in 688d8ff.